feat: Add plugin slots for progress page components

xitij2000 opened the pull request:
Adds a slot for different components in the progress tab to allow them to be overridden with custom components.

The main aim here is to allow a client to support enabling/disabling individual components on a per-course basis. If this is a feature that will be valuable to the broader community we can change the implementation here to directly support that as well. 

This change currently allows overriding individual components. If this is too granular we can also look into creating a slot for the entire page, however that will require more complexity and drift for the overriding component if it just wants to make minor changes. 

![Screenshot 2024-12-12 at 13-07-19 Progress Demonstration Course Nightly](https://github.com/user-attachments/assets/77a3b17f-539d-45f0-a798-b04bb825a255)

openedx-webhooks commented:
Thanks for the pull request, @xitij2000!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/committers-frontend-app-learning`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

xitij2000 commented:
This PR currently doesn't include any screenshots. I will add them before the PR is final depending on the direction of the review.

itsjeyd commented:
@xitij2000 I marked this PR as ready for review assuming that it doesn't change any existing user-facing behavior. Let me know if that's wrong, please (it would need to go through product review in that case).

CC @openedx/committers-frontend-app-learning

bradenmacdonald commented:
@xitij2000 This makes sense to me at a high level. Would you mind adding some screenshots, as you suggested?

@arbrandes Do we have any specific guidelines around where we'll accept slots in the UI, or just use my best judgement?

itsjeyd commented:
@xitij2000 @bradenmacdonald I'm a little unclear on the current status of this PR.

Are there any blockers to starting engineering review?

bradenmacdonald commented:
@itsjeyd I don't think so. I just hoping to hear from @arbrandes.

itsjeyd commented:
@bradenmacdonald Do you want to explicitly request a review from him, maybe? That might help with getting this PR unstuck.

CC @xitij2000

bradenmacdonald commented:
@brian-smith-tcril Could I get your thoughts on this PR? I think it sounds like a good idea, and Adolfo suggested we might want to backport it to Sumac, which I agree with.

brian-smith-tcril commented on src/course-home/progress-tab/ProgressTab.jsx:
```diff
@@ -39,18 +34,15 @@ const ProgressTab = () => {
         <div className="col-12 col-md-4 p-0 px-md-4">
-          {wideScreen && <CertificateStatus />}
-          <RelatedLinks />
+          {wideScreen && <ProgressTabCertificateStatusSlot courseId={courseId} />}
```
I'd like to hear others thoughts on having the `ProgressTabCertificateStatusSlot` in 2 places here. The way this is written now would require plugin authors utilizing the `ProgressTabCertificateStatusSlot` to account for the slot being in the side panel or main body depending on the width of the viewport.

I see a few options here:
* **A**: Keep the slots as written in this PR
  * Requires plugin authors to account for slots being in different places based on viewport width
* **B**: Split `ProgressTabCertificateStatusSlot` into `ProgressTabCertificateStatusMainBodySlot` and `ProgressTabCertificateStatusSidePanelSlot`
  * Requires site operators to put a plugin in 2 slots instead of 1 (even if it's the same plugin)
* **C**: Switch to just having a `MainBody` slot and a `SidePanel` slot instead of granular slots for each component in the body/panel
  * This would remove quite a bit of flexibility in customization
* **D**: Some combination of **(A** ***or*** **B)** ***and*** **C**
  * This would mean keeping the granular slots, but also wrapping those slots in a bigger slot - similar to how `frontend-component-header` has a `DesktopHeaderSlot` that wraps the entire desktop header, while the desktop header component contains more granular slots such as `DesktopUserMenuSlot`
  
  I think I lean towards **B**, with a possibility of adding bigger slots later if needed. I am, however, very open to other opinions on the matter, and would love to hear other options I haven't thought of that others have.

bradenmacdonald replied:
B seems reasonable to me.

xitij2000 replied:
Perhaps there is another option, which is to pass the slot position to component?

i.e. one of the props can be the "placement" of the content, which could be "mainbody" or "sidebar". 

I imagine anyone creating a plugin for this pair of slots would either use the same component or have a single component dynamically adjust to the placement. By explicitly providing this value we can ensure that the plugin author will follow the same logic as the app even if that changes over time. 

Otherwise, I agree that the two slot approach looks good, and I will adapt the code accordingly. I'll push the approach using the "placement" prop since it's a small change, but will switch to the two component approach if that seems better.

brian-smith-tcril reviewed (commented):
Thank you so much for this PR!

Overall I think these are great extension points to have, and I'm super happy to see these documented with screenshots and everything!

I left a few comments with questions and suggestions. Most are quite small, but there's one larger comment that might spark conversation.

xitij2000 commented on src/plugin-slots/ProgressTabRelatedLinksSlot/index.jsx:
```diff
@@ -0,0 +1,20 @@
+const ProgressTabRelatedLinksSlot = ({ courseId }) => (
+  <PluginSlot
+    id="progress_tab_related_links_slot"
+    pluginProps={{
+      courseId,
+    }}
```
I agree. I will remove them from here since the courseId should be unambiguous in these slots and can probably be fetched from the store.

bradenmacdonald replied:
I will eventually want to refactor every page like this in the Learning MFE to not use redux and to use React Context + React Query instead. So if you do need course ID, I would prefer plugins _don't_ fetch it from the store, because that's implicitly making the Redux store part of the API contract here. Our redux stores tend to be messy and buggy because they don't even have a contract _within_ each MFE (they aren't typed using TypeScript), and I really don't want them to become part of the plugin contract, whether officially or unofficially.

In general, my personal view is that plugin slots should pass any props that come from the URL, so that they know exactly which page they're on (in this case it would be the course ID), and any other data that they need should be loaded via small usages of React Query (which should be de-duplicated if any other places on the page need the same data).

In this case, the default content of the plugin slot

> ![Screenshot 2024-11-25 at 10 07 38 AM](https://github.com/user-attachments/assets/02744d24-d852-4729-93b7-f0b06a848a03)

does need the course ID (e.g. "Dates" links to `/learning/course/:courseId/dates`), so I think it makes sense to make that course ID available to any other plugins filling the slot.

xitij2000 replied:
I think perhaps it makes sense to standardise some kind of common context across all MFEs then and plugin can fetch data from there without needing props. I think courseId / learningContextId probably belongs in the context as it is by definition the context in which each component is rendered. At the bare minimum I think the shared common context should have the `contextId`, `username` etc. If the username is missing then a user isn't logged in and if the contextId is missing then it doesn't apply (for instance in the profile page or learner dashboard). 

Or it could be as simple as having hooks like `useContextId` which can then get it from the store right now and from the shared context or whatever mechanism we have later. It seems like too common a usecase here. 

For now, I guess I'll add back the courseId since nothing like the above exists, but it would be a good time now that we have plugins to have a standard way to get common things like this.

bradenmacdonald replied:
> I think perhaps it makes sense to standardise some kind of common context across all MFEs then and plugin can fetch data from there without needing props. I think courseId / learningContextId probably belongs in the context as it is by definition the context in which each component is rendered.

Totally agree with this.

> Or it could be as simple as having hooks like useContextId which can then get it from the store right now and from the shared context or whatever mechanism we have later. It seems like too common a usecase here.

Great idea. Maybe just call it "useLearningContextId" since "Context" is ambiguous to me.

xitij2000 commented:
@brian-smith-tcril I've updated the slots to remove the courseId. I've updated the examples to pull the course ID from the state so that I wouldn't have to update the images though. I hope that's okay!

brian-smith-tcril commented on src/plugin-slots/ProgressTabCertificateStatusSlot/index.jsx:
```diff
@@ -0,0 +1,20 @@
+);
+
+ProgressTabCertificateStatusSlot.propTypes = {
+  placement: PropTypes.oneOf(['MAIN_BODY', 'SIDEBAR']),
```
I'm not sure how I feel about having `placement` as a `pluginProp`.

I think 2 slots provides more flexibility. With 2 slots, site operators can use different `PLUGIN_OPERATIONS` for each, set `keepDefault` differently for each, or set a different `priority` for each.

With `placement` as a `pluginProp` that flexibility is lost.

That being said, I don't know if that flexibility is something we *want*. My current feeling is, "yes, we should provide the flexibility that comes with using 2 slots" - but I'm open to other opinions on the matter.

xitij2000 commented:
I've updated the PR to use two components and moved the logic of hiding or showing the component to the slots so that a plugin can choose to use different logic.

xitij2000 commented:
@brian-smith-tcril  I've fixed the test issues, and the commitlint issues. Thanks for your thorough review!

I feel starting with fewer props and adding more later is less disruptive so keeping courseId out of props seems good for now, and if things change, it will be easy to add it back without hurting existing plugins.

brian-smith-tcril reviewed (approved):
> I feel starting with fewer props and adding more later is less disruptive

I 100% agree!

Thanks for all the work on this! It turned out great!

xitij2000 commented:
> Are there screenshots and/or a description of this PR that I can share with a product manager or instructional designer?
> 
> We have a backlog of pedagogical issues with the progress page, and I'm wondering if we can leverage the pluginslots to get the desired behavior.

This doesn't change the UI at all, so it's hard to screenshot as-is, but I'm adding a screenshot that highlights all the elements that can now be altered with slots.

itsjeyd commented:
@bradenmacdonald Did you want to give this another look before merging?

If not, it looks like the changes will be ready to go after another rebase.

bradenmacdonald commented:
@itsjeyd Nope, this is good to merge without further review from me. Thanks.

itsjeyd commented:
OK, thanks @bradenmacdonald.

@xitij2000 Over to you for a final rebase :)
